### PR TITLE
New code generator

### PR DESCRIPTION
**Pull request by amolenaar**
### PR Type
What kind of change does this PR introduce?

- [ ] Bug fix
- [x] Feature
- [x] Chore (refactoring, formatting, local variables, other cleanup)
- [ ] Documentation content changes

### What is the current behavior?

* Our current code generator is in a state where it's almost impossible to change.
* Our current code generator uses its own parser to parse the model. We can use the `storage` module and `ElementFactory` for that.
* Our current code generator is not a nice example of how to write a code generator

Issue Number: #452 

### What is the new behavior?

* All models are generated with one code generator
* The code generator reads the model using the `storage` module
* The code generator is aware of "super" models (e.g. SysML depends on Core and UML)
* The derived unions can be extended over models
* Attributes and enumerations are defined directly in the class. No separate assignment line
* Black and isort have been disabled in generated files. Some Flake8 warnings are silenced. 
  This ensures the coder output passes pre-commit without need for (manual) change.

### Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### Other information

This PR contains quite a bit of code. The code can be divided in x portions:

1. The code generator (`gaphor.codegen.*`)
2. The generated models: UML, Core, C4Model, SysML, and RAAML (in commit order)
3. Regression tests (`tests/test_models_up_to_date.py`)

Future updates:
* A separate `ModelingLanguage` for the core model. Currently all Core classes are included in the UML model.

### CODE REVIEW

**Comment by amolenaar** [comment]
Fixed by Gaphas 3.4.1

File: tests/test_models_up_to_date.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+"""Test if models are up to date."""
```
Comment (danyeaw) [inline tests/test_models_up_to_date.py:1]: Nice idea to add these tests! :+1:

**Review by danyeaw — COMMENTED** [review]
Nice job removing over 1300 lines of code and cleaning this up! I really like that the profiles can directly depend on the other data models now.

File: gaphor/codegen/coder.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+
+def is_simple_type(c: UML.Class) -> bool:
+    for s in UML.model.get_applied_stereotypes(c):
+        if s.name == "SimpleAttribute":
+            return True
+    for g in c.generalization:
+        if is_simple_type(g.general):
+            return True
+    return False
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py:327]: Nice one :+1:

File: gaphor/codegen/coder.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+    """.format(
+        "isort", "flake8"
+    )  # work around tooling triggers
+)
+
+
+def main(
+    modelfile: str,
+    supermodelfiles: list[tuple[str, str]] = [],
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py:65]: I've been tinkering with this a bit. The default value is `None` now.

File: gaphor/codegen/coder.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+            or is_simple_type(a.type)
+            or is_enumeration(a.type)
+            or is_extension_end(a)
+        ):
+            continue
+        for slot in a.appliedStereotype[:].slot:
+            if slot.definingFeature.name == "subsets":
+                full_name = f"{c.name}.{a.name}"
+                for value in slot.value.split(","):
+                    # TODO: also find derived class in super_models
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py:233]: This TODO has been implemented :see_no_evil:

File: gaphor/codegen/coder.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+    seen_classes = set()
+
+    def order(c):
+        if c not in seen_classes:
+            for b in bases(c):
+                yield from order(b)
+            yield c
+            seen_classes.add(c)
+
+    for c in classes:  # sorted(classes, key=lambda c: c.name):  # type: ignore
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py:302]: Nope. Funny flake8 does not catch this.

File: gaphor/codegen/coder.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+    parser.add_argument(
+        "-o", dest="outfile", type=Path, help="Python data model filename"
+    )
+    parser.add_argument("-r", dest="overridesfile", type=Path, help="Override filename")
+    parser.add_argument(
+        "-s",
+        dest="supermodelfiles",
+        type=str,
+        action="append",
+        help="Reference to dependenct model file (e.g. gaphor.UML.uml:models/UML.gaphor)",
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py:462]: fixed

File: tests/test_models_up_to_date.py (hunk 1 above)
Comment (amolenaar) [reply tests/test_models_up_to_date.py:1]: Indeed, they are our safety net to ensure our models and generated code are in sync.

**Comment by amolenaar** [comment]
> Nice job removing over 1300 lines of code and cleaning this up! I really like that the profiles can directly depend on the other data models now.

I'm also really pleased with how it came about. I really like the model dependencies too.

File: gaphor/codegen/coder.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
+"""The code generator for modeling languages.
+
+This is the code generator for the models used by Gaphor.
+
+In order to work with the code generator, a model should follow some convensions:
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py]: fixed

File: gaphor/codegen/coder.py [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+        ):
+            continue
+        elif redefines(a):
+            redefinitions.append(
+                f'{full_name} = redefine({c.name}, "{a.name}", {a.type.name}, {redefines(a)})'
+            )
+        elif a.isDerived:
+            yield f'{full_name} = derivedunion("{a.name}", {a.type.name}{lower(a)}{upper(a)})'
+        else:
+            if not a.name:
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/codegen/coder.py]: Indeed. Fixed.

**Comment by amolenaar** [comment]
@danyeaw I addressed the items you mentioned.

**Pull request merged**